src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -62,6 +66,171 @@ public static Document Current
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            TransactionManager.Instance.EnsureInTransaction(this.InternalDocument);
+            
+            purgedElementIds.AddRange(PurgeElements(performanceAdviserRuleId, deepPurge));
+            purgedElementIds.AddRange(PurgeMaterials());
+
+            TransactionManager.Instance.TransactionTaskDone();
+
+            if (purgedElementIds.Count == 0)
+                throw new InvalidOperationException(Properties.Resources.NoElementsToPurge);
+
+            return purgedElementIds.Select(x => x.IntegerValue).ToList();
+        }
+
+        private List<ElementId> PurgeMaterials()
+        {
+            List<Autodesk.Revit.DB.ElementId> materials = new FilteredElementCollector(this.InternalDocument).OfClass(typeof(Autodesk.Revit.DB.Material))
+                                                                                                           .ToElementIds()
+                                                                                                           .ToList();
+
+            List<Autodesk.Revit.DB.Element> elementTypes = new FilteredElementCollector(this.InternalDocument).WhereElementIsElementType()
+                                                                                                              .ToElements()
+                                                                                                              .ToList();
+            var nonPurgedElements = new List<Autodesk.Revit.DB.ElementId>();
+
+            for (int i = 0; i < elementTypes.Count; i++)
+            {
+                HostObjAttributes hostObj = elementTypes[i] as HostObjAttributes;
+                if (hostObj != null)
+                {
+                    var compundStructure = hostObj.GetCompoundStructure();
+                    if (compundStructure == null)
+                        continue;
+
+                    int layerCount = compundStructure.LayerCount;
+                    for (int j = 0; j < layerCount; j++)
+                    {
+                        ElementId elementId = compundStructure.GetMaterialId(j);
+                        if (materials.Contains(elementId))
+                        {
+                            nonPurgedElements.Add(elementId);
+                            materials.Remove(elementId);
+                        }
+                        
+                    }
+                    continue;
+                }
+
+                List<ElementId> elementMaterialIds = elementTypes[i].GetMaterialIds(false).ToList();
+                if (elementMaterialIds.Any())
+                {
+                    materials.RemoveAll(purgedId => elementMaterialIds.Exists(elemId => purgedId == elemId));
+                    nonPurgedElements.AddRange(elementMaterialIds.Where(elmeMatId => !nonPurgedElements.Any(nonPurgedElemeMatId => nonPurgedElemeMatId == elmeMatId)));
+                }
+
+                if (materials == null)
```

Comment:
`materials` won't be null ever at this point. if it was null, accessing it above would've thrown already.
did you mean to check if the list is empty ? then do a `if(!materials.Any()) break;` 

if you do want to check it's not `null`, do that right after it's initialised, bypassing a lot of potentially redundant & unsafe (likely to throw) operations